internal/resolution/entitysources/catalogdsource.go

```diff
@@ -71,70 +72,117 @@ func (es *CatalogdEntitySource) Iterate(ctx context.Context, fn input.IteratorFu
+		catalogEntitiesList, err := MetadataToEntities(catalog.Name, channels, bundles)
 		if err != nil {
 			return nil, err
 		}
-		props[entities.PropertyBundlePath] = string(imgValue)
-		catalogScopedPkgName := fmt.Sprintf("%s-%s", bundle.Spec.Catalog.Name, bundle.Spec.Package)
-		bundlePkg := packageMetdatas[catalogScopedPkgName]
-		for _, ch := range bundlePkg.Spec.Channels {
-			for _, b := range ch.Entries {
-				catalogScopedEntryName := fmt.Sprintf("%s-%s", bundle.Spec.Catalog.Name, b.Name)
-				if catalogScopedEntryName == bundle.Name {
-					channelValue, _ := json.Marshal(property.Channel{ChannelName: ch.Name, Priority: 0})
-					props[property.TypeChannel] = string(channelValue)
-					replacesValue, _ := json.Marshal(entities.ChannelEntry{
-						Name:     b.Name,
-						Replaces: b.Replaces,
-					})
-					props[entities.PropertyBundleChannelEntry] = string(replacesValue)
-					entity := input.Entity{
-						ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", bundle.Name, bundle.Spec.Package, ch.Name)),
-						Properties: props,
-					}
-					entityList = append(entityList, entity)
+
+		allEntitiesList = append(allEntitiesList, catalogEntitiesList...)
+	}
+
+	return allEntitiesList, nil
+}
+
+func MetadataToEntities(catalogName string, channels []declcfg.Channel, bundles []declcfg.Bundle) (input.EntityList, error) {
+	entityList := input.EntityList{}
+
+	bundlesMap := map[string]*declcfg.Bundle{}
+	for i := range bundles {
+		bundlesMap[bundles[i].Name] = &bundles[i]
+	}
+
+	for _, ch := range channels {
+		for _, chEntry := range ch.Entries {
+			bundle, ok := bundlesMap[chEntry.Name]
+			if !ok {
+				return nil, fmt.Errorf("bundle %q not found in catalog %q (package %q, channel %q)", chEntry.Name, catalogName, ch.Package, ch.Name)
+			}
+
+			props := map[string]string{}
+
+			for _, prop := range bundle.Properties {
+				switch prop.Type {
+				case property.TypePackage:
+					// this is already a json marshalled object, so it doesn't need to be marshalled
+					// like the other ones
+					props[property.TypePackage] = string(prop.Value)
+				case entities.PropertyBundleMediaType:
+					props[entities.PropertyBundleMediaType] = string(prop.Value)
 				}
 			}
+
+			imgValue, err := json.Marshal(bundle.Image)
+			if err != nil {
+				return nil, err
+			}
+			props[entities.PropertyBundlePath] = string(imgValue)
+
+			channelValue, _ := json.Marshal(property.Channel{ChannelName: ch.Name, Priority: 0})
+			props[property.TypeChannel] = string(channelValue)
+			replacesValue, _ := json.Marshal(entities.ChannelEntry{
+				Name:     bundle.Name,
+				Replaces: chEntry.Replaces,
+			})
+			props[entities.PropertyBundleChannelEntry] = string(replacesValue)
+
+			catalogScopedEntryName := fmt.Sprintf("%s-%s", catalogName, bundle.Name)
+			entity := input.Entity{
+				ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", catalogScopedEntryName, bundle.Package, ch.Name)),
+				Properties: props,
+			}
```

Comment:
The construction of this entity feels quite complex. Is there a particular reason it needs to be? I'm wondering if there is a better way to produce these entities such that there is one for each package, channel, and bundle without having to do this channel+bundle combination stuff.

i.e could the entity creation loop be simplified to something like:
```go
entityList := []input.Entity{}
for _, catalogMetadata := range catalogMetadataList {
  var entity input.Entity
  switch catalogMetadata.Schema {
    case "olm.package":
      //  doing some psuedocode here
      pack := CatalogMetadataToPackage(catalogMetadata)
      entity = input.Entity{
        ID: deppy.IdentifierFromString(IdForPackage(pack)),
        Properties: PropertiesForPackage(pack),
      }
    ...
  }
}
```

Reply:
I'm also not very knowledgeable on the resolution process and what the entity requirements are at this point so please do let me know if something like ^ is just not possible at the moment

Reply:
That's fair. My bad for not properly reading the description and jumping straight to the review :sweat_smile: